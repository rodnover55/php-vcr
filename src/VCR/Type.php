<?php

namespace VCR;

class Type
{
    const HTTP = 'http';
    const PDO = 'pdo';
}
