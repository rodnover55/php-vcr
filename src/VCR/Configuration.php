<?php

namespace VCR;

use VCR\Util\Assertion;

/**
 * Configuration stores a Videorecorders configuration options.
 *
 * Those configuration options might be:
 *  - which LibraryHook to use,
 *  - where to store cassettes or
 *  - which files to scan when filtering source code.
 */
class Configuration
{
    /**
     * @var string Path where cassette files should be stored.
     */
    private $cassettePath = 'tests/fixtures';

    /**
     * List of enabled LibraryHook names.
     *
     * A value of null means all hooks are enabled.
     *
     * @see \VCR\LibraryHooks\LibraryHook
     * @var array List of enabled LibraryHook names.
     */
    private $enabledLibraryHooks;

    /**
     * List of library hooks.
     *
     * Format:
     * array(
     *  'name' => 'class name'
     * )
     * @var array List of library hooks.
     */
    private $availableLibraryHooks = array(
        'stream_wrapper' => 'VCR\LibraryHooks\StreamWrapperHook',
        'curl'           => 'VCR\LibraryHooks\CurlHook',
        'soap'           => 'VCR\LibraryHooks\SoapHook',
    );

    /**
     * Name of the enabled storage.
     *
     * Only one storage can be enabled at a time.
     * By default YAML is enabled.
     *
     * @var string Enabled storage name.
     */
    private $enabledStorage = 'yaml';

    /**
     * List of enabled storages.
     *
     * Format:
     * array(
     *  'name' => 'class name'
     * )
     *
     * @var array List of available storages.
     */
    private $availableStorages = array(
        'blackhole' => 'VCR\Storage\Blackhole',
        'json'      => 'VCR\Storage\Json',
        'yaml'      => 'VCR\Storage\Yaml',
    );

    /**
     * A value of null means all RequestMatchers are enabled.
     *
     * @var array Names of the RequestMatchers which are enabled.
     */
    private $enabledRequestMatchers;

    /**
     * Format:
     * array(
     *  'type => array(
     *      'name' => callback
     *   )
     * )
     *
     * The RequestMatcher callback takes two Request objects and
     * returns true if they match or false otherwise.
     *
     * @var array List of RequestMatcher names and callbacks.
     */
    private $availableRequestMatchers = array(
        Type::HTTP => array(
            'method'       => array('VCR\Drivers\Http\Matcher', 'matchMethod'),
            'url'          => array('VCR\Drivers\Http\Matcher', 'matchUrl'),
            'host'         => array('VCR\Drivers\Http\Matcher', 'matchHost'),
            'headers'      => array('VCR\Drivers\Http\Matcher', 'matchHeaders'),
            'body'         => array('VCR\Drivers\Http\Matcher', 'matchBody'),
            'post_fields'  => array('VCR\Drivers\Http\Matcher', 'matchPostFields'),
            'query_string' => array('VCR\Drivers\Http\Matcher', 'matchQueryString'),
        )
    );

    /**
     * All resource can be in two form:
     * - short: [resource => <class name>] (in this case is called class static method `fromArray`)
     * - extended: [resource => ['class' => <class name>, 'creator' => <callable returns class name>]
    */
    private $factories = array(
        Type::HTTP => array(
            'request' => 'VCR\Request',
            'response' => 'VCR\Response',
            'client' => 'VCR\Drivers\Http\Client'
        )
    );

    /**
     * A whitelist is a list of paths.
     *
     * When processing files for code transformation, only files matching
     * those paths are considered. By default the whitelist is empty and
     * all files (which are not blacklisted) are being considered for
     * code transformation.
     *
     * @var array A whitelist is a list of paths.
     */
    private $whiteList = array();

    /**
     * A blacklist is a list of paths which may not be processed for code transformation.
     *
     * Files in this path are left as is. Blacklisting PHP-VCRs own paths is necessary
     * to avoid infinite loops.
     *
     * @var array A blacklist is a list of paths.
     */
    private $blackList = array('src/VCR/LibraryHooks/', 'src/VCR/Util/SoapClient', 'tests/VCR/Filter');

    /**
     * The mode which determines how requests are handled. One of the MODE constants.
     *
     * @var string Current mode
     */
    private $mode = VCR::MODE_NEW_EPISODES;

    /**
     * List of available modes.
     *
     * Format:
     * array(
     *  'name'
     * )
     *
     * @var array List of available modes.
     */
    private $availableModes = array(
        VCR::MODE_NEW_EPISODES,
        VCR::MODE_ONCE,
        VCR::MODE_NONE,
    );

    /**
     * Returns the current blacklist.
     *
     * @return array
     */
    public function getBlackList()
    {
        return $this->blackList;
    }

    /**
     * Sets one or more paths to blacklist.
     *
     * @param string|string[] $paths Path(s) to blacklist.
     *
     * @return Configuration
     */
    public function setBlackList($paths)
    {
        $paths = (is_array($paths)) ? $paths : array($paths);

        $this->blackList = $paths;

        return $this;
    }

    /**
     * Returns the path to where cassettes are stored.
     *
     * @return string Path to where cassettes are stored.
     */
    public function getCassettePath()
    {
        $this->assertValidCassettePath($this->cassettePath);

        return $this->cassettePath;
    }

    /**
     * Sets the cassette path where a cassettes should be stored in.
     *
     * @param string $cassettePath Path where to store cassettes.
     *
     * @return Configuration
     * @throws VCRException If provided cassette path is invalid.
     */
    public function setCassettePath($cassettePath)
    {
        $this->assertValidCassettePath($cassettePath);
        $this->cassettePath = $cassettePath;

        return $this;
    }

    /**
     * Returns a list of enabled LibraryHook class names.
     *
     * Only class names are returned, any object creation happens
     * in the VCRFactory.
     *
     * @return string[] List of LibraryHook class names.
     */
    public function getLibraryHooks()
    {
        if (is_null($this->enabledLibraryHooks)) {
            return array_values($this->availableLibraryHooks);
        }

        return array_values(array_intersect_key(
            $this->availableLibraryHooks,
            array_flip($this->enabledLibraryHooks)
        ));
    }

    /**
     * Enables specified LibraryHook(s) by its name.
     *
     * @param string|string[] $hooks Name of the LibraryHook(s) to enable.
     * @return Configuration
     * @throws \InvalidArgumentException If a specified library hook doesn't exist.
     */
    public function enableLibraryHooks($hooks)
    {
        $hooks = is_array($hooks) ? $hooks : array($hooks);
        $invalidHooks = array_diff($hooks, array_keys($this->availableLibraryHooks));
        if ($invalidHooks) {
            throw new \InvalidArgumentException("Library hooks don't exist: " . join(', ', $invalidHooks));
        }
        $this->enabledLibraryHooks = $hooks;

        return $this;
    }

    /**
     * Returns the class name of the storage to use.
     *
     * Objects are created in the VCRFactory.
     *
     * @return string Class name of the storage to use.
     */
    public function getStorage()
    {
        return $this->availableStorages[$this->enabledStorage];
    }

    /**
     * Returns a list of enabled http RequestMatcher callbacks.
     *
     * @return array List of enabled RequestMatcher callbacks.
     *
     * @throws \Assert\AssertionFailedException
     */
    public function getRequestMatchers()
    {
        return $this->getTypedRequestMatchers(Type::HTTP);
    }

    /**
     * Returns a list of enabled RequestMatcher callbacks by type.
     *
     * @param $type
     *
     * @return array List of enabled RequestMatcher callbacks.
     *
     * @throws \Assert\AssertionFailedException
     */
    public function getTypedRequestMatchers($type)
    {
        Assertion::minLength($type, 1, "A type of request matchers name must be at least one character long. Found '{$type}'");

        if (is_null($this->enabledRequestMatchers[$type])) {
            return array_values($this->availableRequestMatchers[$type]);
        }

        return array_values(array_intersect_key(
            $this->availableRequestMatchers[$type],
            array_flip($this->enabledRequestMatchers[$type])
        ));
    }

    /**
     * Adds a new http RequestMatcher callback.
     *
     * @param string $name Name of the RequestMatcher.
     * @param callable $callback A callback taking two Request objects as parameters and returns true if those match.
     *
     * @return Configuration
     * @throws VCRException If specified parameters are invalid.
     */
    public function addRequestMatcher($name, $callback)
    {
        return $this->addTypedRequestMatcher(Type::HTTP, $name, $callback);
    }

    /**
     * Adds a new RequestMatcher callback.
     *
     * @param string $type Type of RequestMatcher.
     * @param string $name Name of the RequestMatcher.
     * @param callable $callback A callback taking two Request objects as parameters and returns true if those match.
     *
     * @return Configuration
     *
     * @throws \Assert\AssertionFailedException
     */
    public function addTypedRequestMatcher($type, $name, $callback)
    {
        Assertion::minLength($type, 1, "A type of request matchers name must be at least one character long. Found '{$type}'");
        Assertion::minLength($name, 1, "A request matchers name must be at least one character long. Found '{$name}'");
        Assertion::isCallable($callback, "Request matcher '{$name}' is not callable.");

        $this->availableRequestMatchers[$type][$name] = $callback;

        return $this;
    }

    /**
     * Enables specified http RequestMatchers by its name.
     *
     * @param array $matchers List of RequestMatcher names to enable.
     *
     * @return Configuration
     *
     * @throws \Assert\AssertionFailedException
     */
    public function enableRequestMatchers(array $matchers)
    {
        return $this->enableTypedRequestMatchers(Type::HTTP, $matchers);
    }

    /**
     * Enables specified RequestMatchers by its name and type.
     *
     * @param string $type Type of RequestMatcher
     * @param array $matchers List of RequestMatcher names to enable.
     *
     * @return Configuration
     *
     * @throws \Assert\AssertionFailedException
     */
    public function enableTypedRequestMatchers($type, array $matchers)
    {
        Assertion::minLength($type, 1, "A type of request matchers name must be at least one character long. Found '{$type}'");

        $invalidMatchers = array_diff($matchers, array_keys($this->availableRequestMatchers[$type]));
        if ($invalidMatchers) {
            throw new \InvalidArgumentException("Request matchers don't exist: " . join(', ', $invalidMatchers));
        }
        $this->enabledRequestMatchers[$type] = $matchers;
        
        return $this;
    }

    /**
     * Enables a storage by name.
     *
     * @param string $storageName Name of the storage to enable.
     *
     * @return $this
     *
     * @throws \Assert\AssertionFailedException
     */
    public function setStorage($storageName)
    {
        Assertion::keyExists($this->availableStorages, $storageName, "Storage '{$storageName}' not available.");
        $this->enabledStorage = $storageName;

        return $this;
    }

    /**
      * Returns a list of whitelisted paths.
      *
      * @return array
      */
    public function getWhiteList()
    {
        return $this->whiteList;
    }

    /**
     * Sets a list of paths to whitelist when processing in the StreamProcessor.
     *
     * @param string|array $paths Single path or list of path which are whitelisted.
     *
     * @return Configuration
     */
    public function setWhiteList($paths)
    {
        $paths = (is_array($paths)) ? $paths : array($paths);

        $this->whiteList = $paths;

        return $this;
    }

    /**
      * Returns the current mode.
      *
      * @return string
      */
    public function getMode()
    {
        return $this->mode;
    }

    /**
     * Sets the current mode.
     *
     * @param string $mode The mode to set VCR to
     *
     * @return Configuration
     */
    public function setMode($mode)
    {
        Assertion::choice($mode, $this->availableModes, "Mode '{$mode}' does not exist.");
        $this->mode = $mode;

        return $this;
    }

    /**
     * @return array
     */
    public function getFactories()
    {
        $factories = array();

        foreach ($this->factories as $type => $factory) {
            foreach ($factory as $resource => $definition) {

                $factories[$type][$resource] = is_array($definition) ? $definition : array(
                    'class' => $definition,
                    'creator' => "{$definition}::fromArray"
                );
            }
        }
        return $factories;
    }

    /**
     * @param array $factories
     *
     * @return Configuration
     */
    public function setFactories($factories)
    {
        $this->factories = $factories;

        return $this;
    }

    /**
     * Validates a specified cassette path.
     *
     * @param string $cassettePath Path to a cassette.
     *
     * @throws \Assert\AssertionFailedException
     */
    private function assertValidCassettePath($cassettePath)
    {
        Assertion::directory(
            $cassettePath,
            "Cassette path '{$cassettePath}' is not a directory. Please either "
            . 'create it or set a different cassette path using '
            . "\\VCR\\VCR::configure()->setCassettePath('directory')."
        );
    }
}
