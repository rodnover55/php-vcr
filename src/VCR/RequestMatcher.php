<?php

namespace VCR;

/**
 * @deprecated use \VCR\Drivers\Http\Matcher
 */
class RequestMatcher
{
    /**
     * Returns true if the method of both specified requests match.
     *
     * @param  Request $first  First request to match.
     * @param  Request $second Second request to match.
     *
     * @return boolean True if the method of both specified requests match.
     */
    public static function matchMethod(Request $first, Request $second)
    {
        return $first->getMethod() == $second->getMethod();
    }

    /**
     * Returns true if the url of both specified requests match.
     *
     * @param  Request $first  First request to match.
     * @param  Request $second Second request to match.
     *
     * @return boolean True if the url of both specified requests match.
     */
    public static function matchUrl(Request $first, Request $second)
    {
        return $first->getPath() === $second->getPath();
    }

    /**
     * Returns true if the host of both specified requests match.
     *
     * @param  Request $first  First request to match.
     * @param  Request $second Second request to match.
     *
     * @return boolean True if the host of both specified requests match.
     */
    public static function matchHost(Request $first, Request $second)
    {
        return $first->getHost() === $second->getHost();
    }

    /**
     * Returns true if the headers of both specified requests match.
     *
     * @param  Request $first  First request to match.
     * @param  Request $second Second request to match.
     *
     * @return boolean True if the headers of both specified requests match.
     */
    public static function matchHeaders(Request $first, Request $second)
    {
        // Use array_filter to ignore headers which are null.

        return array_filter($first->getHeaders()) === array_filter($second->getHeaders());
    }

    /**
     * Returns true if the body of both specified requests match.
     *
     * @param  Request $first  First request to match.
     * @param  Request $second Second request to match.
     *
     * @return boolean True if the body of both specified requests match.
     */
    public static function matchBody(Request $first, Request $second)
    {
        return $first->getBody() === $second->getBody();
    }

    /**
     * Returns true if the post fields of both specified requests match.
     *
     * @param  Request $first  First request to match.
     * @param  Request $second Second request to match.
     *
     * @return boolean True if the post fields of both specified requests match.
     */
    public static function matchPostFields(Request $first, Request $second)
    {
        return $first->getPostFields() === $second->getPostFields();
    }

    /**
     * Returns true if the query string of both specified requests match.
     *
     * @param  Request $first  First request to match.
     * @param  Request $second Second request to match.
     *
     * @return boolean True if the query string of both specified requests match.
     */
    public static function matchQueryString(Request $first, Request $second)
    {
        return $first->getQuery() === $second->getQuery();
    }
}
