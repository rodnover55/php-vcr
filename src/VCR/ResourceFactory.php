<?php

namespace VCR;

use VCR\Interfaces\Request;
use VCR\Interfaces\Response;

class ResourceFactory
{
    const DEFAULT_TYPE = Type::HTTP;

    /** @var Configuration */
    private $config;

    public function __construct(Configuration $config)
    {
        $this->config = $config;
    }

    /**
     * @param array $request
     *
     * @return Request
     */
    public function makeRequest(array $request)
    {
        return $this->make('request', $request);
    }

    /**
     * @param array $response
     * @return Response
     */
    public function makeResponse(array $response)
    {
        return $this->make('response', $response);
    }

    public function makeClient($instance)
    {
        $type = $this->getType($instance);

        return $this->make('client', array(
            'type' => $type
        ));
    }

    protected function make($resource, array $data)
    {
        $type = isset($data['type']) ? $data['type'] : self::DEFAULT_TYPE;

        $factories = $this->getFactories();

        if (!isset($factories[$type])) {
            throw new \InvalidArgumentException("Request type '{$type}' doesn't exists");
        }

        $definition = $factories[$type][$resource];

        return call_user_func($definition['creator'], $data);
    }

    public function getType($instance)
    {
        foreach ($this->getFactories() as $type => $resources) {
            foreach ($resources as $definition) {
                if (is_a($instance, $definition['class'])) {
                    return $type;
                }
            }
        }

        $class = get_class($instance);

        throw new \RuntimeException("Type for instance '{$class}' not found");

    }

    protected function getFactories() {
        return $this->config->getFactories();
    }
}
