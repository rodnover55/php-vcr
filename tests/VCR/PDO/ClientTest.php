<?php

namespace VCR\PDO;

use VCR\Drivers\PDO\Client;
use VCR\Drivers\PDO\PDO;
use VCR\Drivers\PDO\Request;

class ClientTest extends TestCase
{
    /** @var Client */
    private $client;

    private $connection = array(
        'dsn' => 'sqlite::memory:',
        'username' => null,
        'password' => null,
        'options' => null
    );

    protected function setUp()
    {
        parent::setUp();

        $this->client = new Client();

        $connection = new PDO($this->connection['dsn']);

        Client::register($connection);
    }

    /**
     * @dataProvider requestsProvider
     *
     * @param Request $request
     */
    public function testSend(Request $request)
    {
        $this->assertEqualsSnapshot($this->client->send($request)->toArray());
    }

    public function requestsProvider()
    {
        return [
            'query' => [
                new Request($this->connection, 'query', 'select 1 as test', [
                    'options' => [
                        'mode' => 19,
                        'object' => null,
                        'ctorargs' => []
                    ]
                ])
            ],
            'exec' => [
                new Request($this->connection, 'exec', 'create table test (id int)')
            ],
            'execPreparedArgs' => [
                new Request($this->connection, 'prepared', 'select ? as test', [
                    'options' => null,
                    'bindings' => [1]
                ])
            ],
            'execPrepared' => [
                new Request($this->connection, 'prepared', 'select ? as test', [
                    'options' => null,
                    'bindings' => [1 => [21, \PDO::PARAM_INT]]
                ])
            ]
        ];
    }
}
