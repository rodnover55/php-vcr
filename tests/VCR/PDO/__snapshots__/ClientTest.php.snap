{
    "testSend with data set \"query\"-0": {
        "type": "pdo",
        "method": "query",
        "result": [
            {
                "test": "1",
                "0": "1"
            }
        ],
        "error": {
            "info": [
                "00000",
                null,
                null
            ]
        }
    },
    "testSend with data set \"exec\"-0": {
        "type": "pdo",
        "method": "exec",
        "result": 0,
        "error": {
            "info": [
                "00000",
                null,
                null
            ]
        }
    },
    "testSend with data set \"execPrepared\"-0": {
        "type": "pdo",
        "method": "prepared",
        "result": [
            {
                "test": "1",
                "0": "1"
            }
        ],
        "error": {
            "info": [
                "00000",
                null,
                null
            ]
        }
    }
}